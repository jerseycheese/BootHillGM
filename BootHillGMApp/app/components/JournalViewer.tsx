import React, { memo, useMemo } from 'react';
import type { JournalEntry } from '../types/journal';
import { isNarrativeEntry, isCombatEntry, isInventoryEntry, isQuestEntry } from '../types/journal';

interface JournalViewerProps {
  entries: JournalEntry[];
}

const JournalEntry = memo(({ entry, formatDate }: { 
  entry: JournalEntry;
  formatDate: (timestamp: number) => string;
}) => {
  const formattedContent = useMemo(() => {
    if (isNarrativeEntry(entry)) {
      return entry.narrativeSummary || entry.content;
    } else if (isCombatEntry(entry)) {
      const content = entry.content.replace(/^\d{1,2}\/\d{1,2}\/\d{4}:\s*/, '').trim();
      return `${formatDate(entry.timestamp)}: ${content}`;
    } else if (isInventoryEntry(entry)) {
      const acquired = entry.items.acquired.length ? `Acquired: ${entry.items.acquired.join(', ')}` : '';
      const removed = entry.items.removed.length ? `Removed: ${entry.items.removed.join(', ')}` : '';
      return [acquired, removed].filter(Boolean).join(' | ');
    } else if (isQuestEntry(entry)) {
      return `${entry.questTitle}: ${entry.status}`;
    }
    return 'Unknown Entry Type';
  }, [entry]);

  return (
    <li className="wireframe-text mb-2">
      <strong>{formatDate(entry.timestamp)}</strong>: {formattedContent}
    </li>
  );
});

JournalEntry.displayName = 'JournalEntry';

const JournalViewer: React.FC<JournalViewerProps> = ({ entries }) => {
  const sortedEntries = useMemo(() => 
    [...entries].sort((a, b) => b.timestamp - a.timestamp), 
    [entries]
  );

const formatDate = (timestamp: number): string => {
    try {
      if (!timestamp || isNaN(timestamp)) {
        console.warn('Invalid timestamp received:', timestamp);
        return 'Invalid date';
      }

      const date = new Date(timestamp); // Fix: create date object
      if (isNaN(date.getTime())) {
        console.warn('Invalid date object created from timestamp:', timestamp);
        return 'Invalid date';
      }

      return new Intl.DateTimeFormat('en-US', {
        month: '2-digit',
        day: '2-digit',
        year: 'numeric'
      }).format(date);
    } catch (error) {
      console.error('Error formatting date:', error);
      return 'Invalid date';
    }
  };

  if (!Array.isArray(entries) || entries.length === 0) {
    return (
      <div className="wireframe-section">
        <h2 className="wireframe-subtitle mb-2">Journal</h2>
        <p>No journal entries yet.</p>
      </div>
    );
  }

  return (
    <div className="wireframe-section">
      <h2 className="wireframe-subtitle mb-2">Journal</h2>
      <ul className="overflow-y-auto">
        {sortedEntries.map((entry) => (
          <JournalEntry
            key={entry.timestamp}
            entry={entry}
            formatDate={formatDate}
          />
        ))}
      </ul>
    </div>
  );
};

export default JournalViewer;
