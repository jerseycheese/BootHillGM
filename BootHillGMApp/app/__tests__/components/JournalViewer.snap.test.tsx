import React from 'react';
import { render } from '@testing-library/react';
import JournalViewer from '@/components/JournalViewer';
import { mockJournalEntries } from '@/test/fixtures/mockComponents';

// Mock formatDate to ensure consistent snapshots
const originalDate = global.Date;
beforeAll(() => {
  // Mock Date constructor to return a fixed date
  const mockDate = new Date(1648635000000); // March 30, 2022
  global.Date = class extends originalDate {
    constructor(date?: string | number | Date) { // Add optional type
      // If a date is provided, call the original constructor
      // Otherwise, initialize with the mock date's time
      super(date ? date : mockDate.getTime());
    }
    // Optionally override other static methods if needed, e.g., Date.now()
    static now() {
      return mockDate.getTime();
    }
// eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any; // Assert the type for assignment to global.Date
});

afterAll(() => {
  global.Date = originalDate;
});

describe('JournalViewer snapshots', () => {
  it('matches snapshot with empty entries', () => {
    const { container } = render(<JournalViewer entries={[]} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with narrative entry', () => {
    const { container } = render(<JournalViewer entries={[mockJournalEntries.narrative]} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with combat entry', () => {
    const { container } = render(<JournalViewer entries={[mockJournalEntries.combat]} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with inventory entry', () => {
    const { container } = render(<JournalViewer entries={[mockJournalEntries.inventory]} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with quest entry', () => {
    const { container } = render(<JournalViewer entries={[mockJournalEntries.quest]} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with multiple entries of different types', () => {
    const allEntries = [
      mockJournalEntries.narrative,
      mockJournalEntries.combat,
      mockJournalEntries.inventory,
      mockJournalEntries.quest,
    ];
    const { container } = render(<JournalViewer entries={allEntries} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with entries in reverse chronological order', () => {
    // Confirm the sort function works by reversing the order
    const reversedEntries = [
      mockJournalEntries.quest,
      mockJournalEntries.inventory,
      mockJournalEntries.combat,
      mockJournalEntries.narrative,
    ];
    const { container } = render(<JournalViewer entries={reversedEntries} />);
    expect(container).toMatchSnapshot();
  });

  it('renders with accessible attributes', () => {
    const { getByTestId } = render(<JournalViewer entries={[mockJournalEntries.narrative]} />);
    const journalViewer = getByTestId('journal-viewer');
    expect(journalViewer).toBeInTheDocument();
    expect(journalViewer).toMatchSnapshot();
  });
});
