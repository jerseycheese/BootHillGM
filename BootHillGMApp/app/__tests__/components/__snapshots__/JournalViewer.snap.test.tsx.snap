// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JournalViewer snapshots matches snapshot with combat entry 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Defeated a bandit
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with empty entries 1`] = `
<div>
  <div
    class="wireframe-section"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <p>
      No journal entries yet.
    </p>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with entries in reverse chronological order 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        The Missing Sheriff: started
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Acquired: Whiskey, Ammunition | Removed: Bandage
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Defeated a bandit
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Arrived at Redemption
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with inventory entry 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Acquired: Whiskey, Ammunition | Removed: Bandage
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with multiple entries of different types 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        The Missing Sheriff: started
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Acquired: Whiskey, Ammunition | Removed: Bandage
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Defeated a bandit
      </li>
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Arrived at Redemption
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with narrative entry 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        Arrived at Redemption
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots matches snapshot with quest entry 1`] = `
<div>
  <div
    class="wireframe-section bhgm-journal-viewer"
    data-testid="journal-viewer"
    id="bhgmJournalViewer"
  >
    <h2
      class="wireframe-subtitle mb-2"
    >
      Journal
    </h2>
    <ul
      class="overflow-y-auto"
    >
      <li
        class="wireframe-text mb-2"
      >
        <strong>
          03/30/2022
        </strong>
        : 
        The Missing Sheriff: started
      </li>
    </ul>
  </div>
</div>
`;

exports[`JournalViewer snapshots renders with accessible attributes 1`] = `
<div
  class="wireframe-section bhgm-journal-viewer"
  data-testid="journal-viewer"
  id="bhgmJournalViewer"
>
  <h2
    class="wireframe-subtitle mb-2"
  >
    Journal
  </h2>
  <ul
    class="overflow-y-auto"
  >
    <li
      class="wireframe-text mb-2"
    >
      <strong>
        03/30/2022
      </strong>
      : 
      Arrived at Redemption
    </li>
  </ul>
</div>
`;
